### ISSUE TITLE
Belos FGMRES solver does not allow missing right preconditioner

### ISSUE THREAD

**Issue by aprokop** [trilinos/Trilinos#4128]
@trilinos/belos @hkthorn 

## Description
I'm trying to run Belos within a nonlinear JFNK solve. As the operator is not constant, it needs FGMRES. I don't have an implemented preconditioner for it, so running without. However, right now it throws due to the check 
```c++
if (isFlexible_) {
  TEUCHOS_TEST_FOR_EXCEPTION(problem_->getRightPrec()==Teuchos::null,BlockGmresSolMgrLinearProblemFailure,
  "Belos::BlockGmresSolMgr::solve(): Linear problem does not have a preconditioner required for flexible GMRES, call setRightPrec().");                  
}
```

I'm not sure why that check is there. Removing it seems to run things just fine. My question is if there is something I'm missing here.

**Comment by cgcgcg** [comment]
@aprokop I am not sure I understand, but shouldn't the operator for a given linear solve be constant? And since there is no preconditioner, regular GMRES should do just fine, no?

**Comment by aprokop** [comment]
@mhoemmen OK. But shouldn't it work for situations where that preconditioner is not variable, and is identity as in this case?

**Comment by mhoemmen** [comment]
@aprokop wrote:
> But shouldn't it work for situations where that preconditioner is not variable, and is identity as in this case? I'm just failing to see why this limitation is in place.

With no preconditioner, the algorithm _should_ just reduce to regular GMRES (ignoring nonlinearity in the "matrix" A).  I suppose Belos could be smart enough to fall back to regular GMRES in the "flexible true but no preconditioner" case.  The work-around would be just to turn off the "flexible" option.

**Comment by mhoemmen** [comment]
@aprokop If you really want Anderson acceleration, NOX has an implementation: see `packages/nox/src/NOX_Solver_AndersonAcceleration.H`.

**Comment by hkthorn** [comment]
I agree with @mhoemmen, there is no reason to use FGMRES if there is no right preconditioner set.  Keeping two orthogonal basis, that are the same, is too much overhead.  I can fix the logic such that it falls back to regular GMRES in this case.

**Comment by hkthorn** [comment]
This has been addressed by commit 55f3eb0.  Marking closed.
